src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,665 @@
+			GetDataFromSegment(struct_children_segment, *children[child_count], total_count);
+		}
+
+	} else {
+		auto aggr_vector_data = FlatVector::GetData(result);
+
+		// set values
+		for (idx_t i = 0; i < segment->count; i++) {
+			if (aggr_vector_validity.RowIsValid(total_count + i)) {
+				GetPrimitiveDataValue(segment, result.GetType(), aggr_vector_data, i, total_count + i);
+			}
+		}
+	}
+}
+
+void ListFun::BuildListVector(LinkedList *linked_list, Vector &result, idx_t &initial_total_count) {
+
+	idx_t total_count = initial_total_count;
+	while (linked_list->first_segment) {
+		auto segment = linked_list->first_segment;
+		GetDataFromSegment(segment, result, total_count);
+
+		total_count += segment->count;
+		linked_list->first_segment = segment->next;
+	}
+
+	linked_list->last_segment = nullptr;
+}
+
+void ListFun::InitializeValidities(Vector &vector, idx_t &capacity) {
+
+	auto &validity_mask = FlatVector::Validity(vector);
+	validity_mask.Initialize(capacity);
+
+	if (vector.GetType().id() == LogicalTypeId::LIST) {
+		auto &child_vector = ListVector::GetEntry(vector);
+		InitializeValidities(child_vector, capacity);
+	} else if (vector.GetType().id() == LogicalTypeId::STRUCT) {
+		auto &children = StructVector::GetEntries(vector);
+		for (auto &child : children) {
+			InitializeValidities(*child, capacity);
+		}
+	}
+}
+
 struct ListAggState {
-	Vector *list_vector;
+	LinkedList *linked_list;
+	LogicalType *type;
+	vector<unique_ptr<AllocatedData>> *owning_vector;
 };
 
 struct ListFunction {
 	template <class STATE>
 	static void Initialize(STATE *state) {
-		state->list_vector = nullptr;
+		state->linked_list = nullptr;
+		state->type = nullptr;
+		state->owning_vector = nullptr;
 	}
 
 	template <class STATE>
 	static void Destroy(STATE *state) {
-		if (state->list_vector) {
-			delete state->list_vector;
+		D_ASSERT(state);
+		if (state->linked_list) {
+			delete state->linked_list;
+			state->linked_list = nullptr;
+		}
+		if (state->type) {
+			delete state->type;
+			state->type = nullptr;
+		}
+		if (state->owning_vector) {
+			state->owning_vector->clear();
+			delete state->owning_vector;
+			state->owning_vector = nullptr;
 		}
 	}
 	static bool IgnoreNull() {
 		return false;
 	}
 };
 
-static void ListUpdateFunction(Vector inputs[], AggregateInputData &, idx_t input_count, Vector &state_vector,
-                               idx_t count) {
+static void ListUpdateFunction(Vector inputs[], AggregateInputData &aggr_input_data, idx_t input_count,
+                               Vector &state_vector, idx_t count) {
 	D_ASSERT(input_count == 1);
 
 	auto &input = inputs[0];
 	UnifiedVectorFormat sdata;
 	state_vector.ToUnifiedFormat(count, sdata);
 
-	auto list_vector_type = LogicalType::LIST(input.GetType());
-
 	auto states = (ListAggState **)sdata.data;
 	if (input.GetVectorType() == VectorType::SEQUENCE_VECTOR) {
 		input.Flatten(count);
```

Comment:
Perhaps just always flatten, if a sequence vector is flattened here and a dictionary vector is flattened inside the `AppendRow`?